File: leetcode/1-100/55-jump-game/bench/jump_game.c
/* LeetCode #55 bench harness — C mirror of jump_game.kara (greedy forward max-reach, star).
 * See ../README.md section Benchmarks. Built with `clang -O3`. */
#include <stdio.h>
#include <stdlib.h>

static long can_jump_work(const long *nums, long n) {
    long farthest = 0;
    long i = 0;
    while (i < n) {
        if (i > farthest) {
            return i;
        }
        if (i + nums[i] > farthest) {
            farthest = i + nums[i];
        }
        if (farthest >= n - 1) {
            return i;
        }
        i += 1;
    }
    return i;
}

int main(void) {
    long total = 200000;
    long modulus = 1000000007;
    long n = 1000;

    long *nums = (long *)malloc((size_t)n * sizeof(long));
    for (long a = 0; a < n; a++) {
        nums[a] = 1 + (a % 4);
    }

    long acc = 0;
    for (long k = 0; k < total; k++) {
        nums[k % n] = 1 + (k % 9);
        long ans = can_jump_work(nums, n);
        acc = (acc * 131 + ans) % modulus;
    }

    printf("%ld\n", acc);
    free(nums);
    return 0;
}

File: leetcode/1-100/55-jump-game/bench/jump_game.rs
// LeetCode #55 bench harness — Rust mirror of jump_game.kara (greedy forward max-reach, ★).
// See ../README.md § Benchmarks. Built with `rustc -O`; the bench.sh also builds a
// `-C overflow-checks=on` variant as the equal-safety comparator to Kāra's checked-by-default
// integer semantics.

fn can_jump_work(nums: &[i64], n: i64) -> i64 {
    let mut farthest: i64 = 0;
    let mut i: i64 = 0;
    while i < n {
        if i > farthest {
            return i;
        }
        if i + nums[i as usize] > farthest {
            farthest = i + nums[i as usize];
        }
        if farthest >= n - 1 {
            return i;
        }
        i += 1;
    }
    i
}

fn main() {
    let total: i64 = 200000;
    let modulus: i64 = 1000000007;
    let n: i64 = 1000;

    let mut nums: Vec<i64> = Vec::new();
    let mut a: i64 = 0;
    while a < n {
        nums.push(1 + (a % 4));
        a += 1;
    }

    let mut acc: i64 = 0;
    let mut k: i64 = 0;
    while k < total {
        nums[(k % n) as usize] = 1 + (k % 9);
        let ans = can_jump_work(&nums, n);
        acc = (acc * 131 + ans) % modulus;
        k += 1;
    }

    println!("{}", acc);
}
